File: support_tools/cs/upd_commander_checker/FlatLayerRules.cs
namespace UpdCommanderChecker;

internal static class FlatLayerRules
{
    private static readonly HashSet<string> ExcludedDirectories =
    [
        "generated",
        "third_party",
        "vendor",
        "external",
        "build",
    ];

    private static readonly HashSet<string> ApplicationMarkers =
    [
        "app",
        "apps",
        "application",
        "applications",
        "feature",
        "features",
    ];

    private static readonly HashSet<string> LayerNames = ["ui", "process", "data"];

    internal static IEnumerable<Finding> Check(
        (
            IReadOnlyList<string> Files,
            string Root,
            IReadOnlyList<IgnoreRule> IgnoreRules,
            int MinFiles,
            int MinDirectPercent
        ) input
    )
    {
        var counts = new Dictionary<string, (int Total, int Direct)>(StringComparer.Ordinal);
        foreach (var file in input.Files)
        {
            var relative = Path.GetRelativePath(input.Root, file).Replace('\\', '/');
            var parts = relative.Split('/', StringSplitOptions.RemoveEmptyEntries);
            if (
                parts
                    .Take(Math.Max(0, parts.Length - 1))
                    .Select(part => part.ToLowerInvariant())
                    .Any(ExcludedDirectories.Contains)
            )
            {
                continue;
            }

            var layerIndex = LayerRootIndex(parts);
            if (layerIndex < 0)
            {
                continue;
            }
            var layerRoot = string.Join('/', parts.Take(layerIndex + 1));
            counts.TryGetValue(layerRoot, out var count);
            count.Total++;
            if (parts.Length == layerIndex + 2)
            {
                count.Direct++;
            }
            counts[layerRoot] = count;
        }

        foreach (var (layerRoot, count) in counts)
        {
            if (
                count.Total < input.MinFiles
                || count.Direct * 100 < count.Total * input.MinDirectPercent
                || IgnoreRules.IsIgnored(
                    new IgnoreCheckInput(layerRoot, "UPD405", string.Empty, input.IgnoreRules)
                )
            )
            {
                continue;
            }
            yield return new Finding(
                layerRoot,
                1,
                "UPD405",
                "large flat layer reduces navigability; consider grouping related responsibilities",
                "attention"
            );
        }
    }

    private static int LayerRootIndex(IReadOnlyList<string> parts)
    {
        var scopeStart = 0;
        for (var index = 0; index + 2 < parts.Count; index++)
        {
            if (ApplicationMarkers.Contains(parts[index].ToLowerInvariant()))
            {
                scopeStart = index + 2;
            }
        }

        var layerIndex = -1;
        for (var index = scopeStart; index + 1 < parts.Count; index++)
        {
            if (LayerNames.Contains(parts[index].ToLowerInvariant()))
            {
                layerIndex = index;
            }
        }
        return layerIndex;
    }
}

File: support_tools/python/upd_commander_checker/src/upd_commander_checker/flat_layer_rules.py
from fnmatch import fnmatch
from pathlib import Path

from .ignore_rules import IgnoreRule
from .models import Finding


_FLAT_LAYER_EXCLUDED_DIRS = {"generated", "third_party", "vendor", "external", "build"}
_APPLICATION_MARKERS = {"app", "apps", "application", "applications", "feature", "features"}
_LAYER_NAMES = {"ui", "process", "data"}


def check_flat_layers(
    paths: list[Path],
    root: Path,
    ignore_rules: tuple[IgnoreRule, ...],
    min_files: int,
    min_direct_percent: int,
) -> list[Finding]:
    counts: dict[str, list[int]] = {}
    for path in paths:
        relative = Path(_relative_text(path, root))
        parts = relative.parts
        lowered = tuple(part.lower() for part in parts)
        if any(part in _FLAT_LAYER_EXCLUDED_DIRS for part in lowered[:-1]):
            continue

        layer_index = _layer_root_index(lowered)
        if layer_index is None:
            continue
        layer_root = Path(*parts[: layer_index + 1]).as_posix()
        direct = len(parts) == layer_index + 2
        bucket = counts.setdefault(layer_root, [0, 0])
        bucket[0] += 1
        if direct:
            bucket[1] += 1

    findings: list[Finding] = []
    for layer_root, (total_files, direct_files) in counts.items():
        if total_files < min_files:
            continue
        if direct_files * 100 < total_files * min_direct_percent:
            continue
        if any(
            rule.code == "UPD405" and fnmatch(layer_root, rule.pattern)
            for rule in ignore_rules
        ):
            continue
        findings.append(
            Finding(
                Path(layer_root),
                1,
                "UPD405",
                "large flat layer reduces navigability; consider grouping related responsibilities",
                "attention",
            )
        )
    return findings


def _layer_root_index(parts: tuple[str, ...]) -> int | None:
    scope_start = 0
    for index in range(len(parts) - 2):
        if parts[index] in _APPLICATION_MARKERS:
            scope_start = index + 2

    layer_index = None
    for index in range(scope_start, len(parts) - 1):
        if parts[index] in _LAYER_NAMES:
            layer_index = index
    return layer_index


def _relative_text(path: Path, root: Path) -> str:
    try:
        return path.relative_to(root).as_posix()
    except ValueError:
        return path.as_posix()
